### PR TITLE
27.10

### PR DESCRIPTION
**Pull request by Winststreloc**
i use my proj, with which i learn web api, i think add automapper for create function

### CODE REVIEW

File: 27.10/PokemonAPI/appsettings.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "ConnectionStrings": {
+    "DefaultConnection": "Data Source=WINST;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"},
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/appsettings.json]: do not commit connection strings 
create your own appsettings.Local.json, add it to gitignore and use Local as env in launchsettings

File: 27.10/PokemonAPI/Data/DataContext.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+﻿using Microsoft.EntityFrameworkCore;
+using PokemonAPI.Models;
+
+namespace PokemonAPI.Data
+{
+    public class DataContext : DbContext
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Data/DataContext.cs]: DbContext name should have 'DbContext' postfix and business domain as prefix
In your case proper name would be PokemonDbContext

**Comment by im-mrogachevskiy** [comment]
Across all application use GUID as ID, never use int

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    public IActionResult GetPokemons()
+    {
+        var pokemons = _mapper.Map<List<PokemonDto>>(_pokemonRepository.GetPokemons());
+        if (!ModelState.IsValid)
+            return BadRequest(ModelState);
+
+        return Ok(pokemons);
+    }
+
+    [HttpGet("{pokemonId}")]
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: pokemonId:int

Always use uri parameter constraints unless it is string

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    [ProducesResponseType(200, Type = typeof(IEnumerable<Pokemon>))]
+    [ProducesResponseType(400)]
+    public IActionResult GetPokemon(int pokemonId)
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: [FromRoute]

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        if (!_pokemonRepository.PokemonExists(pokemonId))
+        {
+            return NotFound();
+        }
+        
+        var pokemon = _mapper.Map<PokemonDto>(_pokemonRepository.GetPokemon(pokemonId));
+        
+        if (!ModelState.IsValid)
+            return BadRequest(ModelState);
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: It is GET, modelState cannot be invalid

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        _pokemonRepository.DeletePokemon(pokemonToDelete);
+        return NoContent();
+    }
+
+    [HttpPost]
+    public IActionResult CreatePokemon([FromQuery] int ownerId,[FromQuery] int categoryId,
+        [FromQuery] int pokemonId,[FromQuery] string pokemonName)
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: all parameters of pokemon should be from body

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        var pokemonCreate = new Pokemon()
+        {
+            Id = pokemonId,
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: Usually we do not get id of new object from client. Either we create in controller or it should be generated by DB

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            return BadRequest(ModelState);
+        }
+
+        _pokemonRepository.CreatePokemon(ownerId, categoryId, pokemonCreate);
+
+        return Ok();
+    }
+
+    [HttpPut]
+    public IActionResult UpdatePokemon([FromQuery] int pokemonId,[FromQuery] string pokemonName)
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: Pass pokemonId as URL parameter. [FromRoute]

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        var pokemonCreate = new Pokemon()
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: fix copypaste

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        var pokemonOwnerEntity = _context.Owners.Where(a => a.Id == ownerId).FirstOrDefault();
+        var category = _context.Categories.Where(a => a.Id == categoryId).FirstOrDefault();
+
+        var pokemonOwner = new PokemonOwner()
+        {
+            Owner = pokemonOwnerEntity,
+            Pokemon = pokemon,
+        };
+
+        _context.Add(pokemonOwner);
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: why do you create any other entities except of pokemon?

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            
+        var pokemonCategory = new PokemonCategory()
+        {
+            Category = category,
+            Pokemon = pokemon,
+        };
+
+        _context.Add(pokemonCategory);
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: same here

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+        _context.Add(pokemon);
+    }
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: _context.SaveChanges()

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+    public void UpdatePokemon(Pokemon pokemon)
+    {
+        _context.Update(pokemon);
+    }
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: _context.SaveChanges()

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+    public bool PokemonExists(int pokemonId)
+    {
+        return _context.Pokemon.Any(p => p.Id == pokemonId);
+    }
+    
+    public void DeletePokemon(Pokemon pokemon)
+    {
+        _context.Remove(pokemon);
+    }
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: _context.SaveChanges()

**Review by im-mrogachevskiy — CHANGES_REQUESTED** [review]

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 15]
```diff
@@ -25,71 +25,57 @@ public PokemonController(IPokemonRepository pokemonRepository, IMapper mapper)
...
-    public IActionResult GetPokemon(int pokemonId)
+    public IActionResult GetPokemon([FromRoute]int pokemonId)
     {
-        if (!_pokemonRepository.PokemonExists(pokemonId))
-        {
-            return NotFound();
-        }
-        
+        if (!_pokemonRepository.PokemonExists(pokemonId)) return NotFound();
+
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: always use {} for 'if' and any other dependent block of code

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 16]
```diff
@@ -25,71 +25,57 @@ public PokemonController(IPokemonRepository pokemonRepository, IMapper mapper)
...
+    public IActionResult DeletePokemon([FromRoute]int pokemonId)
     {
+        if (!_pokemonRepository.PokemonExists(pokemonId)) return NotFound();
         var pokemonToDelete = _pokemonRepository.GetPokemon(pokemonId);
 
-        if (!ModelState.IsValid)
-        {
-            return BadRequest(ModelState);
-        }
+        if (!ModelState.IsValid) return BadRequest(ModelState);
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: what do you validate here?
there's no model to validate

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 17]
```diff
@@ -25,71 +25,57 @@ public PokemonController(IPokemonRepository pokemonRepository, IMapper mapper)
...
 
-        _pokemonRepository.CreatePokemon(ownerId, categoryId, pokemonCreate);
+        var pokemon = _mapper.Map<Pokemon>(pokemonDto);
+        _pokemonRepository.CreatePokemon(ownerId, categoryId, pokemon);
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs]: very good

File: 27.10/PokemonAPI/Controllers/PokemonController.cs [hunk 18]
```diff
@@ -25,71 +25,57 @@ public PokemonController(IPokemonRepository pokemonRepository, IMapper mapper)
...
 
         return Ok();
     }
 
     [HttpPut]
-    public IActionResult UpdatePokemon([FromQuery] int pokemonId,[FromQuery] string pokemonName)
+    public IActionResult UpdatePokemon([FromRoute] int pokemonId, [FromRoute] string pokemonName)
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Controllers/PokemonController.cs:84]: always get data from body
yes, in this case data is just string, but anyway it's better to create new dto class with just this string and use this dto to get data from body

REMEMBER: route is ALWAYS used just to navigate to the recourse, all the data for the resource must be in body

File: 27.10/PokemonAPI/Data/DbContext.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
+﻿using Microsoft.EntityFrameworkCore;
+using PokemonAPI.Models;
+
+namespace PokemonAPI.Data;
+
+public class DbContext : Microsoft.EntityFrameworkCore.DbContext
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Data/DbContext.cs]: PokemonDbContext

File: 27.10/PokemonAPI/Repository/PokemonRepository.cs [hunk 20]
```diff
@@ -27,38 +27,46 @@ public bool PokemonExists(int pokemonId)
...
     public void UpdatePokemon(Pokemon pokemon)
     {
         _context.Update(pokemon);
+        Save();
+    }
+
+    public void Save()
+    {
+        var saved = _context.SaveChanges();
     }
```
Comment (im-mrogachevskiy) [inline 27.10/PokemonAPI/Repository/PokemonRepository.cs]: remove and just use _context.SaveChanges(); everywhere

**Review by im-mrogachevskiy — CHANGES_REQUESTED** [review]
In general, this is getting better
after comments fixed that will be a very good project
I didn't check DB schema and repo but please USE GUIDS
Good job!

**Comment by Sardle** [comment]
You are popusk and daun

**Pull request merged**